common/src/main/java/org/conscrypt/ClientSessionContext.java

```diff
@@ -140,12 +169,10 @@ void onBeforeAddSession(NativeSslSession session) {
         }
 
         HostAndPort key = new HostAndPort(host, port);
-        synchronized (sessionsByHostAndPort) {
-            sessionsByHostAndPort.put(key, session);
-        }
+        putSession(key, session);
 
         // TODO: Do this in a background thread.
-        if (persistentCache != null) {
+        if (persistentCache != null && !session.isSingleUse()) {
```

Comment:
At least for now, we're not going to persistently cache them, yeah.  We currently don't have an interface to selectively delete items from the persistent cache, so if one gets in the persistent cache we might use it over and over.  We can certainly revisit that decision in the future, but I'd rather err on the side of maintaining the single-use property for now.

Reply:
> We currently don't have an interface to selectively delete items from the persistent cache, so if one gets in the persistent cache we might use it over and over.

Hrm. Do you at least delete the ones that expire?

> but I'd rather err on the side of maintaining the single-use property for now.

Fair enough. BTW, QUIC folks found that, in many cases, reading from disk on mobile is slower than the RTT you would otherwise have saved here. So between that and how sensitive the TLS 1.2 sessions are, it might be worth reconsidering persistently caching the 1.2 sessions at all.

Reply:
> Hrm. Do you at least delete the ones that expire?

Not on expiration.  They're kept in a LRUish system that eventually cleans out the old ones, and we only keep one session per host:port pair (overwriting old sessions with new ones).

> Fair enough. BTW, QUIC folks found that, in many cases, reading from disk on mobile is slower than the RTT you would otherwise have saved here. So between that and how sensitive the TLS 1.2 sessions are, it might be worth reconsidering persistently caching the 1.2 sessions at all.

Good to know, that might be a nice simplification.

Reply:
There's a bit of a complication here, in that now the persistent cache will end up with a multi-use ticket in it, that will be used for every new connection (when there isn't a session in the temporary session cache), so if we get a single-use ticket it will never actually remove the ticket from the persistent cache.

If you're maintaining 1.2 caching in the persistent cache, you'll probably want to toss a comment here noting that behavior.

Reply:
Added a comment to the places we retrieve things from the persistent cache that we might find a multi-use session even if we've previously been using single-use sessions.